src/derivative_utils.jl

```diff
@@ -51,6 +51,8 @@ function calc_J(integrator, cache::OrdinaryDiffEqConstantCache, is_compos)
   if DiffEqBase.has_jac(f)
     J = f.jac(uprev, p, t)
   else
+    cache.uf.t = t
+    cache.uf.p = p
```

Comment:
Great! :+1:  Can you remove the updates of `uf` in the oop version of `calc_W!`?

Reply:
I'm thinking of delete oop version of `cache_W!` directly.

Reply:
And only use `update_W!`? Currently I don't like that `calc_W!` is called `calc_W!` both for in-place and out-of-place functions but updates `W` only for the in-place functions and returns a new `W` for the out-of-place functions. So I thought maybe it could be done similar to `calc_J` and `calc_J!`, i.e., we could have an out-of-place variant `calc_W` and an updating variant `calc_W!`.

Moreover, I'm not happy with the fact that at the moment there's a lot of reuse strategies hidden in the implementation of `calc_W!`. Maybe it would be cleaner if `calc_W!` would just calculate `W`, without considering any reuse strategies, similar to what we do in `calc_J`. Ideally then we would have a separate function (such as `update_W!`) that first figures out if `J` should be updated (and then calls `calc_J!`) and then if `W` should be updated (and then calls `calc_W!`).

Reply:
Agree.